src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs

```diff
@@ -298,11 +319,222 @@ void writeEnumElement(string kind, int value, string operationName, string? extr
+                        }
+                    }
+
+                    writeStandardBaseArguments(type switch
+                    {
+                        { IsAbstract: true } => "kind",
+                        { IsInternal: true } => "OperationKind.None",
+                        _ => $"OperationKind.{GetKind(type)}"
+                    });
+                    Outdent();
+                    // The base class is responsible for initializing its own properties,
+                    // so we only initialize our own properties here.
+                    if (type.Properties.Count == 0)
+                    {
+                        // Note: our formatting style is a space here
+                        WriteLine(" { }");
+                    }
+                    else
+                    {
+                        Blank();
+                        Brace();
+                        foreach (var prop in type.Properties)
+                        {
+                            WriteLine($"{prop.Name} = {prop.Name.ToCamelCase()};");
+                        }
+                        Unbrace();
+                    }
+
+                    foreach (var prop in type.Properties)
+                    {
+                        WriteLine($"public {prop.Type} {prop.Name} {{ get; }}");
+                    }
+
+                    if (type is Node node)
+                    {
+
+                        var visitorName = GetVisitorName(node);
+                        WriteLine($@"public override IEnumerable<IOperation> Children => Array.Empty<IOperation>();
+        public override void Accept(OperationVisitor visitor) => visitor.{visitorName}(this);
+        public override TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument) => visitor.{visitorName}(this, argument);");
+                    }
 
-        private bool IsImmutableArray(string typeName)
+                    Unbrace();
+                }
+            }
+
+            WriteLine("#endregion");
+
+            void writeStandardConstructorParameters(bool isAbstract)
+            {
+                if (isAbstract)
+                {
+                    Write("OperationKind kind, ");
+                }
+
+                Write("SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit");
+            }
+
+            void writeStandardBaseArguments(string kind)
+            {
+                Write($"{kind}, semanticModel, syntax, type, constantValue, isImplicit)");
+            }
+        }
+
+        private void WriteVisitors()
         {
-            return typeName.StartsWith("ImmutableArray<", StringComparison.Ordinal);
+            WriteLine("#region Visitors");
+            WriteLine(@"public abstract partial class OperationVisitor
+    {
+        public virtual void Visit(IOperation operation) => operation?.Accept(this);
```

Comment:
similarly, this doesn't really matter if the generated code is well-formed, but it feels like these multiline string literals are awkwardly encoding assumptions about the exact indent depth of the class currently being generated.

Reply:
I thought about this, but my conclusion was that this was better from a readability perspective. If you have significant pushback I can refactor, but this seems like a pretty safe thing to hardcode to me.

---
In reply to: [313615218](https://github.com/dotnet/roslyn/pull/37791#discussion_r313615218) [](ancestors = 313615218)